liferay-frontend/liferay-portal
# LPS-118339 Adds Support and Implementation for Remote Apps JS SDK

This is the mostly-squashed history originally developed here:

-   https://github.com/wincent/liferay-portal/pull/361
-   https://github.com/wincent/liferay-portal/pull/365

### High-level overview

-   Remote apps are hosted on remote servers, and run in iframes on DXP pages.
-   There is a demo remote app running at: http://remote-app-test.wincent.com/ (this is a GitHub pages site that serves static HTML, CSS, JS).
-   Source for that at: https://github.com/wincent/remote-app-test
-   In this PR, we have two new modules:
    - `remote-app-client-js` provides a JS SDK that remote apps use to communicate with their parent DXP instance.
    - `remote-app-support-web` provides the corresponding DXP-side support.

![ScreenFlow](https://user-images.githubusercontent.com/7074/90534196-bc9e3a00-e179-11ea-9510-f666a116b596.gif)

### Capabilities of the SDK in this POC

This POC implements an "MVP" of the structure proposed in [the research doc](https://docs.google.com/document/d/1L3AXKG4k229lN0w-IA0EWzOy-dflSUnGjuv-1E1w2cA/edit); that is, it provides a way for a remote app to:

- Send simple client-side commands from the iframe to the parent DXP window (eg. `openToast`). The list of supported operations can obviously be extended/adjusted.
- Retrieve basic client-side information from the parent (eg. query for things like `isSignedIn`, `userName` etc); I included these items only as examples — we can add to or remove from this list.
- Run `fetch` queries as the currently logged in user via a proxy. That is, remote app running on host `remote.example.com` can tell its parent (where user is logged in) to fetch stuff from `dxp-cloud.example.net`. **An important consequence of this is that remote apps must be trusted:** Even though a remote app cannot run arbritrary JS in the parent frame (due to browser security policy for iframes), the proxy means that it can read/write data as the logged in user of the DXP instance, to the extent that REST APIs and other endpoints. If this is deemed too permissive, we're going to have to adopt a whitelist of of accessible endpoints.
- Run a GraphQL query via the same mechanism. The same security considerations apply.
- Ask for a CSS URL from the host DXP instance so that it can incorporate and override those styles inside its iframe, if it wants. At the moment, this is just the `clay.css` URL from the theme root.

### Limitations

- We're not yet implementing any of the advanced security features discussed in the design document (OAuth2 flow, signed URLs, trusted apps). This avoids a bunch of complexity, but it does limit what remote apps can do on _their_ server sides (eg. they can't do their own REST requests). On the bright side, that also means that we don't yet have to worry about the security and legal (eg. GDPR) implications of sharing access privileges with a third-party environment outside of the user's browser. But eventually we'll have to have a bulletproof policy in place.
- Maybe stating the obvious, but the `fetch()` and `graphQL` proxy APIs are incredibly low level and are not the most developer-friendly of things, as they require the developer to construct valid REST requests and GraphQL queries will few affordances. At least on the GraphQL side developers can introspect the schema; I don't know what nice things we have to help people discover and formulate valid REST requests (presumably something? 🤷).
- The CSS URL feature is a bit gimmicky at the moment because the app can only query for the CSS URL _after_ the SDK has loaded, which means you get a Flash Of Unstyled Content when you load the iframe. If we want this to be a real feature, then we need to pass this kind of information to the remote app via query string parameters in the URL, so that it has a chance to load the CSS from the HTML `<head>` before the body is rendered; and if we do that, there may be other interesting info that we can pass along with the request.
- Speaking of asynchronous stuff, an unavoidable limitation of iframe-based communication is that it is all async. This means that a lot of the methods in the API return promises (except things like `openToast`, which is fire-and-forget). In the event of an error, the promises reject. I added the ability to listen for error events (eg. `on('error', handleError)`) for places where no explicit promise is involved, but in practice it may not be super useful. We'll see if we want to keep the event bus around, but it was one of the things contemplated in the design doc.
- Likewise, for the DXP-side support, I'm loading that using a `BottomDynamicInclude`. In my testing, that means the support code loads *after* the remote app iframe is already running. I implemented graceful handling of that (the SDK retries with exponential backoff until the support side is online), but we may want to way up how eagerly/early we load the support stuff. As it is, I just copied a similar module (`frontend-js-tooltip-support-web`) and used that as a basis.
- Because I was sort of designing this on the fly, I haven't gone crazy with the test coverage nor have I tried to refactor it to be minimal or anything like that — given that it all may change a lot, I prefer to keep it as "dumb" as possible so that it can be easily changed, and only at "the end", do a pass to clean it all up and boost coverage.
- In initial testing I was seeing the iframe load twice on every page load, but at least of today, it appears to be only loading once. Either something got fixed on the backend 👏 or something was weird with my portal instance days ago 🎉. 

### Other observations

-   `remote-app-client-js` uses webpack to bundle the SDK for external consumption. In this PR our demo app is actually hitting DXP to download the SDK, but we can also release it as an NPM package and/or host it on a CDN.
-   Seeing as it is a self-contained and isolated bundle, I took the opportunity to write the code in TypeScript, which helped me catch bugs during development and gave me helpful context (tooltips, autocomplete) along the way. To make this official, we'll need to bake the necessary dev dependencies into liferay-npm-scripts and tweak some settings (for example, to enable TS formatting; for now I formatted manually with `yarn run liferay-npm-scripts prettier --write {src,test}/**/*.ts`). I think it's worth doing this, as it provides us with a small stepping stone towards getting TS into liferay-portal. On the other hand, if we can't or don't want to do this, I can easily strip out all the TS types, rename some files, tweak the webpack config, and turn all this back into vanilla JS.

### Test plan

- Go to "Remote Apps" under "Custom Apps" in the global menu.
- Add a remote app with name "REMOTE APP" and URL http://remote-app-test.wincent.com/
- Back on the site, click on the top nav bar edit icon.
- Search for "REMOTE" and drag the widget to the page.
- Publish the page.
- Open the browser console and click the buttons.

Most of the buttons do exactly what you'd expect. The one to watch out for is the "Trigger Error" button, which messes with the internal state a bit (creating a duplicate client), so be aware that after you've clicked that one, things might not work all that well until the next page reload. 😂
